psi4/src/psi4/libfock/dft_integrators.h

```diff
@@ -218,38 +218,58 @@ inline void rks_gradient_integrator(std::shared_ptr<BasisSet> primary, std::shar
                                     SharedMatrix G, SharedMatrix U, int ansatz = -1) {
     ansatz = (ansatz == -1 ? fworker->ansatz() : ansatz);
 
-    // Get scratch pointers
-    double** Gp = G->pointer();
+    // => Setup scratch pointers, and associated variables <= //
+    auto Gp = G->pointer();
 
-    double** Up = U->pointer();
-    double** Tp = pworker->scratch()[0]->pointer();
-    double** Dp = pworker->D_scratch()[0]->pointer();
+    auto Up = U->pointer();
+    auto Tp = pworker->scratch()[0]->pointer();
+    auto Dp = pworker->D_scratch()[0]->pointer();
 
     // Fine for now, but not true once we start caching
-    int max_functions = U->ncol();
+    auto max_functions = U->ncol();
 
-    // Get block data
-    int npoints = block->npoints();
-    double* w = block->w();
-    const std::vector<int>& function_map = block->functions_local_to_global();
-    int nlocal = function_map.size();
+    // => Per-block setup <= //
+    auto npoints = block->npoints();
+    auto w = block->w();
+    const auto& function_map = block->functions_local_to_global();
+    auto nlocal = function_map.size();
 
-    // Get points data
-    double** phi = pworker->basis_value("PHI")->pointer();
-    double** phi_x = pworker->basis_value("PHI_X")->pointer();
-    double** phi_y = pworker->basis_value("PHI_Y")->pointer();
-    double** phi_z = pworker->basis_value("PHI_Z")->pointer();
-    double* rho_a = pworker->point_value("RHO_A")->pointer();
-    size_t coll_funcs = pworker->basis_value("PHI")->ncol();
+    // => Setup accessors to computed values <= //
+    auto phi = pworker->basis_value("PHI")->pointer();
+    auto phi_x = pworker->basis_value("PHI_X")->pointer();
+    auto phi_y = pworker->basis_value("PHI_Y")->pointer();
+    auto phi_z = pworker->basis_value("PHI_Z")->pointer();
+    auto rho_a = pworker->point_value("RHO_A")->pointer();
+    auto coll_funcs = pworker->basis_value("PHI")->ncol();
 
-    // => LSDA Contribution <= //
-    double* v_rho_a = fworker->value("V_RHO_A")->pointer();
+    // NOTE:
+    // Let d/dx be a derivative of a functon in R^3 with respect to coordinate x.
+    // Let d/d(x,i) be a derivative with respect to displacing atom i in the x direction.
+    // Then d/d(x,i) phi_mu,p = - d/dx phi^x_mu,p * δ_mu,i
+    //    δ = 1 if mu centered on atom i, else 0.
+    // i.e., the change in a basis function at a point if you move right is the same as the
+    // change if you move the basis function left.
+
+    // NOTE:
+    // ∂E/∂D ∂D/∂x = 0 because ∂E/∂D = 0 by Hellmann-Feynman.
+    // Therefore, we don't compute density-derivative terms. They sum to zero.
```

Comment:
This is not Hellmann-Feynman! It's just because we assume that the density has been optimized and the SCF equations are satisfied, which makes the derivative zero.

Reply:
The comments were the thing I wanted a second opinion on. Can you elaborate on how this is not Hellmann-Feynman? We've chosen all "wavefunction" parameters to minimize the energy function, so the "wavefunction" response terms do not appear in the gradient expression. That's how I remember the Hellmann-Feynman theorem being talked about.

Reply:
The derivative `dE/dP` is zero simply because the orbitals are assumed to correspond to an extremum of the total energy <=> the SCF equations are satisfied. IIRC this is how one ends up with the term with the energy-weighted density matrix.

Reply:
Ah. I follow Helgaker in adding a Lagrangian term to enforce orbital orthonormality to the Hamiltonian, so that Hellmann-Feynman still applies, but I concede that isn't a universal convention. I can amend the comment.

Reply:
> Ah. I follow Helgaker in adding a Lagrangian term to enforce orbital orthonormality to the Hamiltonian, so that Hellmann-Feynman still applies, but I concede that isn't a universal convention. I can amend the comment.

I don't follow; that term is always in the Lagrangian.

Still not Hellmann-Feynman.

Reply:
See Sec 5.3 [here](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=e795e93080574f7f9fc769549e526370d0a23442), for example. We apply Hellmann-Feynman to the _Lagrangian_ rather than the original Hamiltonian. Either way, amending the comment now.